keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+
+If a lien is set to Deployment A, only a deletion request to Deployment A is blocked.
+Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.
+
+If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
+Users can still request to delete Deployment A and Pod C directly.
+When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
+  - otherwise: Deployment A is deleted immediately
+- Background cascading deletion: Deployment A is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked until all other resources without liens are deleted.
+This means a resource with lien will eventually be deleted.
+Users can still request to delete its namespace or other objects in the namespace.
+
+TODO: Update below description depending on the decision.
+
+```
+When deletion of the namespace is requested, the resource deletion in the deleted namespace is done by namespace lifecycle controller in below two steps:
+1. Delete all resources (as a result, only resources without lien are deleted)
+2. Force delete all resources (as a result, all resources are deleted regardless of lien)
+
+It is to ensure proper order of deletion of resources as much as possible, even when they are force deleted through namespace deletion.
+```
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
```

Comment:
```suggestion
To avoid such a misunderstanding from users, the API server must block any request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp`.
```

(changing a **should** to a **must**; we could defer implementing that 'til beta if it helps)